libraries/Microsoft.Bot.Builder/TurnContext.cs

```diff
@@ -260,7 +252,7 @@ Task<ResourceResponse[]> SendActivitiesThroughCallbackPipeline(int nextCallbackI
                 // Send from the list which may have been manipulated via the event handlers. 
                 // Note that 'responses' was captured from the root of the call, and will be
                 // returned to the original caller.
-                var responses = await this.Adapter.SendActivities(this, bufferedActivities.ToArray());
+                var responses = await Adapter.SendActivities(this, bufferedActivities.ToArray());
```

Comment:
At the risk of breaking into Performance land, I suspect the "ToArray()" should not be here. Passing in the buffer would let the lower level better optimize the send.

Reply:
humm - i'd have to investigate this further - its a little weird to see these ToArrays

Reply:
I'll spare you the investigation 'cause I already did it. It's totally possible to change this all to `IEnumerable<IActivity>`. In fact I had a local branch that is 99% of the way to `IEnumerable<IActivity>` which I threw out because there was a lot of code churn since I started it and redoing the work is actually trivial. In the code churn, the implementation actually changed and revealed that we would end up buffering the set internally anyway since it needed to be transformed and enumerated more than once, so for me, while I'd like to see the change, it deprioritized it.

Another reason I stopped working on it is actually that in some places we currently need `Activity` not `IActivity` and when I stumbled across this I decided it would be better to have a broader conversation around the fact that we shouldn't even have interfaces rather than just switching the interfaces, but then still hardcoding casts to `Activity` internally where the switch still needed to happen which could then break if someone passed a non `Activity` based `IActivity`. The API this calls, `BotAdapter::SendActivities`, for example requires `Activity[]` right now as well. 

The change isn't hard, it all needs to be aligned... and I don't want to take it on and submit a PR without having a discussion about it first so I don't waste anyone's time.